### ISSUE TITLE
Backups segfault on ARMv7

### ISSUE THREAD

**Issue by richard-scott** [restic/restic#2618]
I build restic with Jenkins from Github master as and when commits are pushed.
Until this time I have not had any issues with it backing files up... until now ;-(

Output of `restic version`
--------------------------
```
# restic version
debug log file debug.log
debug enabled
restic 0.9.6 (v0.9.6-135-ga307797c) compiled with go1.14 on linux/arm
```

How did you run restic exactly?
-------------------------------
```
export B2_ACCOUNT_ID="XXXX"
export B2_ACCOUNT_KEY="XXXX"
export RESTIC_REPOSITORY="b2:restic-$(hostname | md5sum | awk '{print $1}')-$(hostname)"
export RESTIC_PASSWORD_FILE="/root/.restic"
export RESTIC_CACHE_DIR="/data/.restic"
restic init
restic backup /etc
restic backup /etc
```

The 1st backup of /etc will work ok, but the 2nd backup is when things segfault.

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Backblaze B2

Expected behavior
-----------------
Restic should backup correctly.
My last successful backup on this host was on the 27th Feb 2020.
(I have another linux host that bacups hourly and is working just fine.)

Actual behavior
---------------
It dies with a Segmentation fault :-(
```
# restic backup /etc
debug log file debug.log
debug enabled
repository 4357d291 opened successfully, password is correct
Segmentation fault files 4.264 MiB, total 941 files 6.448 MiB, 0 errors ETA 0:47
root@ReadyNAS:~/test#
/etc/wgetrc
```
Each time I run a backup it crashes but It doesn't stop at the same point in time or file.

Steps to reproduce the behavior
-------------------------------
I'm just running a backup (well, attempting to)

Do you have any idea what may have caused this?
-----------------------------------------------
I am using a [ReadyNAS Duo](https://www.netgear.com/support/product/RND2000v1_(ReadyNAS_Duo_v1).aspx) to run this backup on. It has 512MB of RAM, and lots of swap space allocated that it never actually uses.

Do you have an idea how to solve the issue?
-------------------------------------------
Restic backs up just fine on a Linux x86_64 system with 16GB RAM. Maybe its a memory thing, but its odd that its only just stopped working.
I have deleted my repo and started again with a fresh one at Backblaze so I don't believe data to be corrupted.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Yes, and no.

**Comment by fd0** [comment]
Hm, that's very odd. Do you get a Go stack trace? Anything else printed to stderr? What does the kernel log say?

Without knowing too much about your specific situation my first guess is that this is an issue with your system, e.g. a hardware issue. I have extensive experience with the Go runtime, and up to now I never had a segfault without a stack trace...

How do you build restic exactly? Using `build.go`? or manually? Is `cgo` enabled?

**Comment by richard-scott** [comment]
I'm building restic with the following command on my x64_64 system and the scp'ing the binary over to the NAS.
```
go run build.go --goos linux --goarch arm -tags debug
```

Actually that gave me an idea... I've just downloaded the offical [binary ](https://github.com/restic/restic/releases/download/v0.9.6/restic_0.9.6_linux_arm.bz2) and that works!

Maybe my build process is wrong? Is it?

**Comment by MichaelEischer** [comment]
Please try whether compiling the restic master branch with go 1.13 yields a working binary. As another sanity check you could also compile the commit used for the v0.9.6 release with go 1.14.

**Comment by fd0** [comment]
FYI, there's also the environment variable `GOARM`, maybe you need to set that? The official binaries are also built with `build.go` (using an empty `GOARM` value), maybe that's it?

I'm still wondering about the missing output to stderr, what does the kernel log say?

**Comment by richard-scott** [comment]
I've not got GOARM compiled... I've not changed my Jenkins job for months, and its only just started failing in the last 6-8 weeks.
There is nothing in the kernel logs and even looking at repo stats doesn't work for me:

```
$ restic --verbose=1000  stats
debug log file debug.log
debug enabled
repository 4357d291 opened successfully, password is correct
Segmentation fault
$
```

The debug.log ends with this:
```
2020/03/10 16:09:51 repository/index.go:503     repository.DecodeIndex  95      done
2020/03/10 16:09:51 repository/index.go:553     repository.LoadIndexWithDecoder 95      Loading index 06af2d143950ed481718eab4849d9f57680949a5b3da0afbf1458ac61d23ddcd
2020/03/10 16:09:51 repository/repository.go:77 repository.(*Repository).LoadAndDecrypt 95      load index with id 06af2d143950ed481718eab4849d9f57680949a5b3da0afbf1458ac61d23ddcd
2020/03/10 16:09:51 cache/backend.go:164        cache.(*Backend).Load   95      Load(<index/06af2d1439>, 0, 0) from cache
2020/03/10 16:09:51 cache/file.go:44    cache.(*Cache).Load     95      Load from cache: <index/06af2d1439>
2020/03/10 16:09:51 repository/index.go:458     repository.DecodeIndex  95      Start decoding index
2020/03/10 16:09:51 repository/index.go:458     repository.DecodeIndex  92      Start decoding index
2020/03/10 16:09:51 repository/index.go:503     repository.DecodeIndex  92      done
2020/03/10 16:09:51 repository/index.go:503     repository.DecodeIndex  94      done
2020/03/10 16:09:51 repository/index.go:458     repository.DecodeIndex  93      Start decoding index
2020/03/10 16:09:51 repository/index.go:503     repository.DecodeIndex  93      done
2020/03/10 16:09:51 repository/index.go:553     repository.LoadIndexWithDecoder 93      Loading index 082d901fa3836e5890eff2d394a5ed3e89fa4d16e5219574d390629ecde0b78e
2020/03/10 16:09:51 repository/repository.go:77 repository.(*Repository).LoadAndDecrypt 93      load index with id 082d901fa3836e5890eff2d394a5ed3e89fa4d16e5219574d390629ecde0b78e
2020/03/10 16:09:51 cache/backend.go:164        cache.(*Backend).Load   93      Load(<index/082d901fa3>, 0, 0) from cache
2020/03/10 16:09:51 cache/file.go:44    cache.(*Cache).Load     93      Load from cache: <index/082d901fa3>
2020/03/10 16:09:51 repository/index.go:458     repository.DecodeIndex  93      Start decoding index
2020/03/10 16:09:51 repository/index.go:553     repository.LoadIndexWithDecoder 92      Loading index 09b8d107e9ac06de6a4b382af3955c3842bc866a7dacb01a17abd08617fe7eb0
2020/03/10 16:09:51 repository/repository.go:77 repository.(*Repository).LoadAndDecrypt 92      load index with id 09b8d107e9ac06de6a4b382af3955c3842bc866a7dacb01a17abd08617fe7eb0
2020/03/10 16:09:51 cache/backend.go:164        cache.(*Backend).Load   92      Load(<index/09b8d107e9>, 0, 0) from cache
2020/03/10 16:09:51 cache/file.go:44    cache.(*Cache).Load     92      Load from cache: <index/09b8d107e9>
2020/03/10 16:09:51 repository/index.go:458     repository.DecodeIndex  92      Start decoding index
2020/03/10 16:09:51 repository/index.go:503     repository.DecodeIndex  92      done
2020/03/10 16:09:51 repository/index.go:553     repository.LoadIndexWithDecoder 94      Loading index 0a749ef0f3af5c03d065758d31294ade7627c255cb192791b31db94e51ae5126
2020/03/10 16:09:51 repository/repository.go:77 repository.(*Repository).LoadAndDecrypt 94      load index with id 0a749ef0f3af5c03d065758d31294ade7627c255cb192791b31db94e51ae5126
2020/03/10 16:09:51 cache/backend.go:164        cache.(*Backend).Load   94      Load(<index/0a749ef0f3>, 0, 0) from cache
2020/03/10 16:09:51 cache/file.go:44    cache.(*Cache).Load     94      Load from cache: <index/0a749ef0f3>
2020/03/10 16:09:51 repository/index.go:458     repository.DecodeIndex  94      Start decoding index
2020/03/10 16:09:51 repository/index.go:553     repository.LoadIndexWithDecoder 92      Loading index 0a64c4151402f4a873e64f215b416e9af6febb36ce0bef28c617c64f0e59d08d
2020/03/10 16:09:51 repository/repository.go:77 repository.(*Repository).LoadAndDecrypt 92      load index with id 0a64c4151402f4a873e64f215b416e9af6febb36ce0bef28c617c64f0e59d08d
2020/03/10 16:09:51 cache/backend.go:164        cache.(*Backend).Load   92      Load(<index/0a64c41514>, 0, 0) from cache
2020/03/10 16:09:51 cache/file.go:44    cache.(*Cache).Load     92      Load from cache: <index/0a64c41514>
2020/03/10 16:09:51 repository/index.go:458     repository.DecodeIndex  92      Start decoding index
2020/03/10 16:09:51 repository/index.go:503     repository.DecodeIndex  92      done
2020/03/10 16:09:51 repository/index.go:553     repository.LoadIndexWithDecoder 92      Loading index 0b6ed9aec76a902d0a1bf510b50709e91a2f6452e775e072f0924f8a229fe643
2020/03/10 16:09:51 repository/repository.go:77 repository.(*Repository).LoadAndDecrypt 92      load index with id 0b6ed9aec76a902d0a1bf510b50709e91a2f6452e775e072f0924f8a229fe643
2020/03/10 16:09:51 cache/backend.go:164        cache.(*Backend).Load   92      Load(<index/0b6ed9aec7>, 0, 0) from cache
2020/03/10 16:09:51 cache/file.go:44    cache.(*Cache).Load     92      Load from cache: <index/0b6ed9aec7>
```

**Comment by richard-scott** [comment]
@MichaelEischer - your Go-1.13.4 idea resulted in a good backup:

```
# restic version
debug log file debug.log
debug enabled
restic 0.9.6 (c542a509) compiled with go1.13.4 on linux/arm
#
# restic backup /etc
debug log file debug.log
debug enabled
repository 4357d291 opened successfully, password is correct

Files:           0 new,     1 changed,   940 unmodified
Dirs:            0 new,     0 changed,     0 unmodified
Added to the repo: 350 B

processed 941 files, 6.448 MiB in 1:45
snapshot 22b658f8 saved

```

**Comment by MichaelEischer** [comment]
So that make it more likely to be a problem with Go. Can you could try the following for Go 1.14:
Clean the go build cache `go clean -cache` and then build again and give it another try?

**Comment by expipiplus1** [comment]
I also recently experienced restic segfaulting on armv7l:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x353e8c]

goroutine 208 [running]:
github.com/restic/restic/internal/repository.(*Index).ID(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        internal/repository/index.go:377 +0x3c
github.com/restic/restic/internal/repository.(*Repository).LoadIndex.func5(0x4, 0x8c5e50)
        internal/repository/repository.go:467 +0x88
golang.org/x/sync/errgroup.(*Group).Go.func1(0x1348f90, 0x1336ef0)
        vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x48
created by golang.org/x/sync/errgroup.(*Group).Go
        vendor/golang.org/x/sync/errgroup/errgroup.go:54 +0x50
```

`restic version`: `restic 0.9.5 compiled with go1.12.9 on linux/arm`

I realize this was built with an older go, and this *seems* to have been a one time thing so I'm not too concerned. I just thought I should include this here for completeness.

**Comment by richard-scott** [comment]
I can compile successfully with go v1.14, but I can't run an incremental backup.
I can do the initial backup, list snapshots etc etc, but not an incremental as it bombs out loading the index.

**Comment by MichaelEischer** [comment]
> `restic version`: `restic 0.9.5 compiled with go1.12.9 on linux/arm`

That crash is unrelated to the segfault with go 1.14 and has been fixed in restic 0.9.6 via #2321 .

@richard-scott That looks like a bug in go 1.14. Can you try whether setting the environment variable `GODEBUG=asyncpreemptoff=1` help? You could also try to add `-gcflags="all=-N"` as compilation flag, which disables the new defer optimizations.

**Comment by expipiplus1** [comment]
Ah great, thanks!

**Comment by greatroar** [comment]
Confirmed on a Raspberry Pi. Backups work sometimes, but restic stats consistently segfaults. Running this in gdb points to [this line](https://github.com/golang/go/blob/go1.14/src/runtime/signal_unix.go#L432) as the culprit.

**Comment by richard-scott** [comment]
This:
```
$ GODEBUG=asyncpreemptoff=1 restic backup /etc 
```

didn't help.

Do you mean the other flags to be used like this:
```
$ go run -gcflags=all=-N build.go --goos linux --goarch arm
```

If so, that didn't help either ;-(

**Comment by greatroar** [comment]
Turns out that what I'm seeing is just the OOM killer. @richard-scott can you try

```
echo 2 > /proc/sys/vm/overcommit_memory
```

then run restic again?

**Comment by greatroar** [comment]
I think I found it. restic/progress installs signal handlers with [signal.Notify](https://godoc.org/os/signal#Notify) in various places, but uses unbuffered channels. When I make a few of the channels buffered, and I have overcommit set to 2, the SIGSEGV is properly caught by the Go runtime and I get

```
fatal error: runtime: out of memory
```

followed by a stack trace.

**Comment by MichaelEischer** [comment]
@richard-scott Could you test whether #2644 or Go 1.14.1 solves your crash?

**Comment by MichaelEischer** [comment]
@richard-scott Could you test whether #2644 or Go 1.14.1 solves your crash?

**Comment by richard-scott** [comment]
@MichaelEischer I'm affraid to say that Go 1.14.2 didn't help, neither did the #2644 built from [here](https://github.com/greatroar/restic/tree/signal-notify-buffered).

@greatroar - neither did this help:

```
echo 2 > /proc/sys/vm/overcommit_memory
```

I can create a repo on B2 just fine, I can backup the 11Mb from /etc, but then I can't add new backups of /etc to this even if nothing has changed locally.

**Comment by richard-scott** [comment]
This version: restic_0.9.6_linux_arm works just fine BTW.

**Comment by greatroar** [comment]
Strange. It might be that the channels still don't have "sufficient buffer space to keep up with the expected signal rate", or something else entirely may be the cause of this.

(#2644 will still make these problems easier to debug.)

**Comment by greatroar** [comment]
It looks like golang.org/x/crypto's ARM assembler has become broken due to changes in the Go runtime. That was [fixed](golang/crypto/commit/61a87790db17894570dfb32dbaa0a4af9ce60cb4) in January.

@richard-scott Can you try building restic after doing

```
go get -u golang.org/x/crypto
```

in the source directory?

**Comment by richard-scott** [comment]
I get this output:

```
$ go get -u golang.org/x/crypto
go: golang.org/x/crypto upgrade => v0.0.0-20200429183012-4b2356b1ed79
go: downloading golang.org/x/crypto v0.0.0-20200429183012-4b2356b1ed79
```

And I can now start a backup with the latest build from master.
Fantastic work guys, thank you so much :-)
